File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.java
import java.util.*;

class Solution {
    public double[] solution(int k, int[][] ranges) {
        List<Integer> a = new ArrayList<>();
        a.add(k);
        
        while (k > 1) {
            k = k % 2 == 0 ? k / 2 : 3 * k + 1;
            a.add(k);
        }
        
        double[] area = new double[a.size()];
        for (int i = 0; i< a.size() - 1; i++) {
            area[i + 1] = area[i] + (a.get(i) + a.get(i + 1)) / 2.0;
        }
        
        int n = area.length - 1;
        double[] answer = new double[ranges.length];
        
        for (int i = 0; i < ranges.length; i++) {
            int[] r = ranges[i];
            int left = r[0];
            int right = n + r[1];
            
            if (left > right) {
                answer[i] = -1;
                continue;
            }
            
            answer[i] = area[right] - area [left];
        }
        return answer;
    }
}

File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.py
def solution(k, ranges):
    
    nums = [k]
    
    while k > 1:
        if k % 2 == 0:
            k //= 2
        else:
            k *= 3
            k += 1
        nums.append(k)
    
    answer = []
    for a, b in ranges:
        b += len(nums) - 1    
        if a > b:
            answer.append(-1)
            continue

        result = 0
        for i in range(a, b):
            result += (nums[i] + nums[i+1]) / 2
        answer.append(result)
        
    return answer